streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -104,14 +105,18 @@
         this.parentDir = parentDir;
     }
 
+
     @SuppressWarnings("unchecked")
     public void openDB(final ProcessorContext context) {
         // initialize the default rocksdb options
+
+        options = new Options();
+
         final BlockBasedTableConfig tableConfig = new BlockBasedTableConfig();
         tableConfig.setBlockCacheSize(BLOCK_CACHE_SIZE);
         tableConfig.setBlockSize(BLOCK_SIZE);
+        tableConfig.setFilter(new BloomFilter());
```

Comment:
Hi @bbejeck,

Should we also consider `options.optimizeFiltersForHits()` to save memory on the bloom filter in exchange for one I/O for each get on a missing key?

(see https://github.com/facebook/rocksdb/wiki/Memory-usage-in-RocksDB#indexes-and-filter-blocks)

I just ran across this while reading about caching in Rocks.

Reply:
@vvcephei yeah that makes sense.

\cc @guozhangwang @mjsax  WDYT?

Reply:
FWIW, here's what I'm thinking...
Bloom filters can only tell you if the key is definitely not in the set. They can only answer "no" or "maybe". So if the filter says the key isn't in some sst we don't have to actually do the I/O to check, but if it says "maybe", we do still have to check.

This optimization would add a bloom filter to every level in the SST hierarchy except the last level. 

The rationale is that if you're pretty sure the keys you get are in the db, the bloom filters at the higher levels would let you skip querying the SSTs that don't contain your key. If you get all the way to the bottom level, we're pretty sure the key is there (via our prior assumption), so checking the bloom filter isn't that valuable, since it would rarely answer "no".
If it answers "maybe", we have to check anyway. In other words, the filter only saves I/O in the rare case that it does say "no".
On the other hand, the last level has the most keys in it, so those are the most expensive filters. By dropping that last level of filters, we save a bunch of memory in exchange for rare extra I/Os.

Do we have a prior assumption that the keys we query for are rarely missing? I think so...
In general, Streams only does a get while computing an aggregation value, etc. In this case, it does a get followed by a put. Therefore, the only get that might return missing is the very first one for each key.

Factors that would cause more missed gets would be stuff like:
* IQ
* data sets that have a lot of thrash in the key space

Reply:
Thanks for the clarification @vvcephei sounds good to me.

Reply:
Thanks @vvcephei for the explanation. I agree with you that for most cases we expect multiple gets on each key (only the first get will miss), besides the ones you already listed that may not be the case another case is that windowed stream-stream join will always have distinct keys, but given that this may be fixed in the future I'm in favor of adding it as well.